docs: EXPOSED-747 Add a get started tutorial for Exposed DAO

vnikolova opened the pull request:
#### Description

Add a new "Get started with Exposed's DAO API" tutorial to mirror the "Get started with Exposed" tutorial.

Additionally:
- group both tutorials in an empty "Get started" TOC element.
- add a new snippets project `get-started-with-exposed-dao`.
- include `get-started-with-exposed` to the snippets project.
- rename the `Getting-Started-with-Exposed.topic` file to match the title.

---

#### Type of Change

Please mark the relevant options with an "X":
- [x] Documentation update

---

#### Related Issues
[EXPOSED-747](https://youtrack.jetbrains.com/issue/EXPOSED-747)

bog-walk commented on documentation-website/Writerside/topics/Get-started-with-Exposed-DAO.md:
```diff
@@ -0,0 +1,268 @@
+SQL: INSERT INTO TASKS ("name", DESCRIPTION, COMPLETED) VALUES ('Learn Exposed DAO', 'Follow the DAO tutorial', FALSE)
+SQL: INSERT INTO TASKS ("name", DESCRIPTION, COMPLETED) VALUES ('Read The Hobbit', 'Read chapter one', TRUE)
+Created new tasks with ids 1 and 2
+Completed tasks:
+SQL: SELECT TASKS.ID, TASKS."name", TASKS.DESCRIPTION, TASKS.COMPLETED FROM TASKS WHERE TASKS.COMPLETED = TRUE
```
Based on the code in `App`, the output will actually be different.
Calling `Task.find()` returns a value of type `SizedIterable<Task>`, so when `SizedIterable.count()` is called after, the final SQL that gets sent is something like : `SELECT COUNT(*) ...`.

If the SQL above is the output we do want, then we'd have to chain `Task.find { ... }.toList()` instead to return a value of `List<Task>`.

Depending on how much detail you want to get into this tutorial, I'd suggest the second option. Otherwise, it'd be prudent to add a note about what exactly the return type of `find()` is.

bog-walk commented on documentation-website/Writerside/topics/Get-started-with-Exposed-DAO.md:
```diff
@@ -0,0 +1,268 @@
+
+First, you create the tasks table using the `SchemaUtils.create()` method. The `SchemaUtils` object holds utility 
+methods for creating, altering, and dropping database objects.
+
+Once the table has been created, you use the `IntEntity` extension method `.new()` to add two new `Task` records:
```
Semantics, but might need to change the wording of this to avoid users nit-picking:

`new()` is technically an `IntEntityClass` method (same for `find()` and `all()`). But they can be called on an `IntEntity` class only thanks to the inclusion of the companion object. Which is why they could not be called on an actual `IntEntity` instance, like `task1.find()` wouldn't compile.

bog-walk commented on documentation-website/Writerside/topics/Get-started-with-Exposed-DAO.md:
````diff
@@ -0,0 +1,268 @@
+```
+{src="get-started-with-exposed-dao/src/main/kotlin/org/example/App.kt" include-symbol="task1,task2"}
+
+Within the `new` block, you set the values for each column. Exposed will translate the functions into the following
+SQL queries:
````
Could we also make a short mention of the type of `task1` and `task2` being an actual instance of the `IntEntity` class that was defined? Something like what's done [in the docs](https://www.jetbrains.com/help/exposed/dao-crud-operations.html#create)

bog-walk commented on documentation-website/Writerside/topics/Get-started-with-Exposed-DAO.md:
```diff
@@ -0,0 +1,268 @@
+
+## Update and delete a task
+
+Let’s extend the app’s functionality by updating and deleting the same task.
```
Example below updates and deletes different tasks

bog-walk commented on documentation-website/Writerside/topics/Get-started-with-Exposed-DAO.md:
```diff
@@ -0,0 +1,268 @@
+SQL: SELECT COUNT(*) FROM TASKS WHERE TASKS.COMPLETED = TRUE
+Completed tasks: 1
+SQL: SELECT TASKS.ID, TASKS."name", TASKS.DESCRIPTION, TASKS.COMPLETED FROM TASKS WHERE TASKS.COMPLETED = TRUE
+Task(id=2, title=Read The Hobbit, completed=true)
+Updated task1: Task(id=1, title=Learn Exposed DAO, completed=true)
```
The output for an entity won't look this clean unless we define it ourselves. Since we're not using a data class, an override could just be added, like:
```kt
class Task(id: EntityID<Int>) : IntEntity(id) {
    // ...
    override fun toString(): String {
        return "Task(id=$id, title=$title, completed=$isCompleted)"
    }
}
```

bog-walk commented on documentation-website/Writerside/topics/Get-started-with-Exposed-DAO.md:
````diff
@@ -0,0 +1,268 @@
+```
+{src="get-started-with-exposed-dao/src/main/kotlin/org/example/App.kt" include-lines="11,13-15,34-41"}
+
+You update the value of a property just as you would with any property in a Kotlin class:
+
+```kotlin
+```
+{src="get-started-with-exposed-dao/src/main/kotlin/org/example/App.kt" include-lines="35"}
````
I think it's good to keep this tutorial basic, so this is the only section I'd recommend adding a tad more complexity to.
I'd suggest switching the update to something like this:
```kt
// ...
task1.title = "Try Exposed DAO"
task1.isCompleted = true
println("Updated task1: $task1")
// ...
```
Particularly because the output below will still show only a single generated SQL, emphasising that Exposed caches the changes made to the entity instance and waits until the ideal moment to actually send the (final) changes together to the database:
```
SQL: UPDATE TASKS SET COMPLETED=TRUE, "name"='Try Exposed DAO' WHERE ID = 1
```

Which could be a natural lead to adding a small note section (like [this one in the docs](https://www.jetbrains.com/help/exposed/dao-crud-operations.html#update)).

bog-walk reviewed (changes requested):
Looking good so far, just some request for the code samples

vnikolova commented:
Thanks for the review @bog-walk 🙏 I updated the code examples, added the suggested notes and included the `get-started-with-exposed` project to the root one. Can you please verify the changes?

bog-walk commented on documentation-website/Writerside/snippets/get-started-with-exposed/README.md:
```diff
@@ -1,12 +1,20 @@
 
 A sample Gradle/Kotlin project built by following the steps explained in
-the [Get started with Exposed](https://jetbrains.github.io/Exposed/getting-started-with-exposed.html) tutorial.
+the [Get started with Exposed](https://www.jetbrains.com/help/exposed/getting-started-with-exposed.html) tutorial.
```
Given that the name of the original .topic file has now changed, how will it work when the documentation is later synced? Will this original url path (and the one referenced in the root project README) still work or will they then break? Just curious.

vnikolova replied:
It will still work, because we've set the property `accepts-web-file-names="getting-started-with-exposed.html"` in the `hi.tree` file. Good catch though, the URL should be updated anyway.

bog-walk replied:
Ah got it, thanks!

bog-walk commented on documentation-website/Writerside/topics/Get-Started-with-Exposed.topic:
```diff
@@ -333,8 +336,8 @@
                     file and add the following transaction function:
                 </p>
                 <code-block lang="kotlin"
-                            src="get-started-with-exposed/app/src/main/kotlin/org/example/App.kt"
-                            include-lines="4-22,28-45,60-61"
+                            src="get-started-with-exposed/src/main/kotlin/org/example/App.kt"
+                            include-lines="4-20,26-45,58-60"
```
Not sure if I'm in the right section or if this was a problem before, but I just noticed:

In the 'Create and Query a Table' section of 'Get started with Exposed DSL', the main code block has a weird indentation. Almost like it has an extra parentheses somewhere.

bog-walk commented on documentation-website/Writerside/topics/Get-started-with-Exposed-DAO.md:
```diff
@@ -0,0 +1,280 @@
+SQL: CREATE TABLE IF NOT EXISTS TASKS (ID INT AUTO_INCREMENT PRIMARY KEY, "name" VARCHAR(128) NOT NULL, DESCRIPTION VARCHAR(128) NOT NULL, COMPLETED BOOLEAN DEFAULT FALSE NOT NULL)
+SQL: INSERT INTO TASKS ("name", DESCRIPTION, COMPLETED) VALUES ('Learn Exposed DAO', 'Follow the DAO tutorial', FALSE)
+SQL: INSERT INTO TASKS ("name", DESCRIPTION, COMPLETED) VALUES ('Read The Hobbit', 'Read chapter one', TRUE)
+Created new tasks with ids 1 and 2
+SQL: SELECT COUNT(*) FROM TASKS WHERE TASKS.COMPLETED = TRUE
+Completed tasks: 1
```
Line 250 ' SELECT COUNT(*)...' needs to be changed to match line 209 above.

bog-walk reviewed (approved):
Mentioned a missing output change. Thanks for putting this together!

e5l reviewed (approved):
lgtm

zamulla commented on documentation-website/Writerside/topics/Get-started-with-Exposed-DAO.md:
```diff
@@ -0,0 +1,280 @@
+
+
+<procedure>
+Before you start using Exposed, you need to provide dependencies to your project.
```
```suggestion
Before you start using Exposed, you need to add dependencies to your project.
```

zamulla commented on documentation-website/Writerside/topics/Get-started-with-Exposed-DAO.md:
```diff
@@ -0,0 +1,280 @@
+Before you start using Exposed, you need to provide dependencies to your project.
+<step>
+
+Navigate to the **gradle/libs.versions.toml** file and define the Exposed and H2 version and libraries:
```
```suggestion
Navigate to the **gradle/libs.versions.toml** file and define the Exposed and H2 versions and artifacts:
```

zamulla commented on documentation-website/Writerside/topics/Get-started-with-Exposed-DAO.md:
```diff
@@ -0,0 +1,280 @@
+
+## Configure a database connection
+
+Every database access using Exposed is started by obtaining a connection and creating a transaction.
```
I wouldn't say "access is started", maybe something like "Every time you access a database using Exposed, you start by..."

zamulla commented on documentation-website/Writerside/topics/Get-started-with-Exposed-DAO.md:
```diff
@@ -0,0 +1,280 @@
+name for the table. If you don't provide a name, Exposed will derive it from the object name, which may lead to
+unexpected results depending on naming conventions.
+
+The `Tasks` object defines three columns:
```
It's safer to avoid using exact numbers in this context :) to save ourselves some maintenance down the road, when a column is added and there is no easy way to track if we referred to the number of columns anywhere.

zamulla reviewed (changes requested):
Nice! I've suggested some low-level changes.
